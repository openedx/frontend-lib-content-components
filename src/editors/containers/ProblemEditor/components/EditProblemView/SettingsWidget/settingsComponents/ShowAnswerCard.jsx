import React, { useState, useEffect } from 'react'
import { injectIntl, FormattedMessage } from '@edx/frontend-platform/i18n';
import SettingsOption from '../SettingsOption';
import { Form, Hyperlink } from '@edx/paragon';
import { ShowAnswerTypes } from '../../../../../../data/constants/problem';
import PropTypes from 'prop-types';
import { problemDataProps } from '../../../../../../data/services/cms/types';

export const ShowAnswerCard = ({
    showAnswer,
    updateSettings,
}) => {
    const [summary, setSummary] = useState("")
    const handleAnswerTypeChange = (event) => {
        updateSettings({ showAnswer: { ...showAnswer, on: event.target.value } })
    }
    const handleAttemptChange = (event) => {
        updateSettings({ showAnswer: { ...showAnswer, afterAttempts: parseInt(event.target.value) } })
    }
    const createSummary = () => {
        for (const key in ShowAnswerTypes) {
            if (showAnswer.on === ShowAnswerTypes[key].value) {
                setSummary(ShowAnswerTypes[key].name)
                break
            }
        }
    }

    useEffect(() => {
        createSummary()
    })

    return (
        <SettingsOption title="Show Answer" summary={summary} >
            <div>
                <span>Set attempt logic for when learners can see the correct answer.</span>
            </div>
            <div>
                <Hyperlink destination="#" target="_blank">
                    Set a default value in advanced settings
                </Hyperlink>
            </div>
            <Form.Group>
                <Form.Control
                    as="select"
                    value={showAnswer.on}
                    onChange={handleAnswerTypeChange}
                >
                    {Object.keys(ShowAnswerTypes).map((showAnswerType, i) => (
                        <option
                            key={i}
                            value={ShowAnswerTypes[showAnswerType].value}
                        >
                            {ShowAnswerTypes[showAnswerType].name}
                        </option>
                    ))}
                </Form.Control>
            </Form.Group>
            <Form.Group>
                <Form.Control
                    type="number"
                    value={showAnswer.afterAttempts}
                    onChange={handleAttemptChange}
                    floatingLabel="Number of Attempts"
                />
            </Form.Group>
        </SettingsOption>
    )
}

ShowAnswerCard.propTypes = {
    showAnswer: problemDataProps.settings.showAnswer,
    updateSettings: PropTypes.func.isRequired,
};

export default injectIntl(ShowAnswerCard);