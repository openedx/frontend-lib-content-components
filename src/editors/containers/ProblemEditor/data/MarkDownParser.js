/*
This is the direct motivation code taken from:
https://github.com/open-craft/capa-visual-editor/

Changes have been made to adapt the parser with respect to our use case.
*/

import { ProblemTypeKeys } from "../../../data/constants/problem";

export const indexToLetterMap = [...Array(26)].map((val, i) => String.fromCharCode(i + 65));

function getHints(markdown) {
  if (!markdown || !markdown.trim()) {
    return [];
  }
  const hints = [];

  function getHint(row) {
    let hint = '';
    if (row.startsWith('||')) {
      const hintStart = row.indexOf('||');
      const hintEnd = row.slice(hintStart + 2).indexOf('||');
      hint = row.slice(hintStart + 2, hintEnd + 2);
    }
    return hint;
  }
  for (const d in markdown.split('\n')) {
    const row = markdown.split('\n')[d];
    const hint = getHint(row);
    if (hint.length) {
      hints.push({
        id: Number(d),
        value: hint,
      });
    }
  }
  return hints;
}

function getShortAnswerOptions(markdown) {
  // TODO: Include tolerence for number it fails for now.
  let problemType = ProblemTypeKeys.TEXTINPUT;
    if ((!markdown || !markdown.trim())) {
      return {
        shortAnswersList: [],
        problemType,
      };
    }
    const markdownListData = markdown.split('\n');
    const shortAnswersList = [];
    let gotAnswer = false;
    for (const i in markdownListData) {
      const row = markdownListData[i].trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      let feedback = '';
      let answer = '';
      let correct = true;
      if (row.startsWith('=') && !gotAnswer) {
        gotAnswer = true;
        if (feedbackStart > -1) {
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
          answer = row.slice(1, feedbackStart).trim();
        } else {
          answer = row.slice(1).trim();
        }
      } else if (row.startsWith('=') && gotAnswer) {
        break;
      } else if (gotAnswer && row.startsWith('or=')) { // the next correct answer
        if (feedbackStart > -1) {
          answer = row.slice(3, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(3).trim();
        }
      } else if (gotAnswer && row.startsWith('not=')) { // the incorrect answer
        if (feedbackStart > -1) {
          answer = row.slice(4, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(4).trim();
        }
        correct = false;
      }
      if (answer) {
        problemType = isFinite(answer) ? ProblemTypeKeys.NUMERIC : ProblemTypeKeys.TEXTINPUT;
        shortAnswersList.push({
          id: indexToLetterMap[shortAnswersList.length],
          title: answer,
          feedback,
          correct,
        });
      }
    }
    return {
      shortAnswersList,
      problemType,
    };
}

function getMultipleChoiceOptions(markdown) {
  const multipleChoiceOptions = [];
  const data = {
    multiSelectAnswersList: multipleChoiceOptions,
    problemType: ProblemTypeKeys.MULTISELECT,
    groupFeedbackList: [],
  };
  if (!markdown || !markdown.trim()) {
    return data;
  }
  const markdownListData = markdown.split('\n');
  let feedback = '';
  for (const d in markdownListData) {
    const row = markdownListData[d];
    // multiple choices
    if (row.startsWith('[ ]') || row.startsWith('[x]')) {
      // title
      let title = row.slice(row.indexOf(']') + 1).trim();

      const feedbacksStart = row.indexOf('{{');
      const selectedFeedbackStart = row.indexOf(' selected:') || row.indexOf('{selected:'); // two ways to define selected feedback
      const selectedFeedbackEnd = row.slice(selectedFeedbackStart + 1).indexOf('}') + selectedFeedbackStart;
      const unselectedFeedbackStart = row.indexOf('unselected:');
      const unselectedFeedbackEnd = row.slice(unselectedFeedbackStart + 1).indexOf('}}') + unselectedFeedbackStart;

      let selectedFeedback = '';
      let unselectedFeedback = '';
      let titleChanged = false;
      if (feedbacksStart !== -1 && selectedFeedbackStart !== -1) {
        selectedFeedback = row.slice(selectedFeedbackStart + 10, selectedFeedbackEnd);
        title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        titleChanged = true;
      }
      if (feedbacksStart !== -1 && unselectedFeedbackStart !== -1) {
        unselectedFeedback = row.slice(unselectedFeedbackStart + 12, unselectedFeedbackEnd + 1);
        if (!titleChanged) {
          title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        }
      }
      multipleChoiceOptions.push({
        id: indexToLetterMap[multipleChoiceOptions.length],
        title,
        correct: !row.startsWith('[ ]'),
        selectedFeedback,
        unselectedFeedback,
      });
    }
    // group Feedback
    if (multipleChoiceOptions.length) {
      if (row.startsWith('{{')) {
        const groupFeedbackEnd = row.indexOf('}}');
        const goupStart = row.indexOf('((');
        const groupEnd = row.indexOf('))');
        const group = row.slice(goupStart + 2, groupEnd).split(' ');
        const groupChoices = [];
        for (const i in group) {
          const letter = group[i];
          if (multipleChoiceOptions[indexToLetterMap.indexOf(letter)]) {
            groupChoices.push(multipleChoiceOptions[indexToLetterMap.indexOf(letter)].id);
          }
        }
        feedback = row.slice(groupEnd + 2, groupFeedbackEnd).trim();

        if (groupChoices.length && feedback) {
          data.groupFeedbackList.push({
            id: groupChoices.length,
            answers: groupChoices,
            feedback,
          });
        }
      }
    }
  }
  return data;
}

function getSingleChoiceOptions(markdown) {
  // TODO: Build support for selected and unselected feedback
  if (!markdown || !markdown.trim()) {
    return {
      singleSelectAnswersList: [],
      problemType: ProblemTypeKeys.DROPDOWN,
    };
  }

  const singleChoiceOptions = [];
  let dropDownMode = false;
  const markdownListData = markdown.split('\n');
  for (const d in markdownListData) {
    const row = markdownListData[d].trim();
    if (row.startsWith('[[')) {
      dropDownMode = true;
      continue;
    }
    if (row.startsWith(']]')) {
      break;
    }
    if (dropDownMode) {
      if (row.startsWith('(')) {
        let feedback = '';
        let title = row.slice(row.indexOf('(') + 1, row.indexOf(')')).trim();
        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = title.slice(0, feedbackStart).trim();
        }
        const correct = true;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      } else {
        let feedback = '';
        let title = '';
        // let title = row.slice(row.indexOf(')') + 1).trim();

        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = row.slice(0, feedbackStart).trim();
        } else {
          title = row.trim();
        }
        const correct = false;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      }
    } else if (row.startsWith('( )') || row.startsWith('(x)')) {
      let feedback = '';
      let title = row.slice(row.indexOf(')') + 1).trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
        feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
        title = title.slice(0, feedbackStart - 4).trim();
      }
      const correct = !!row.startsWith('(x)');
      singleChoiceOptions.push({
        id: indexToLetterMap[singleChoiceOptions.length],
        title,
        correct,
        feedback,
      });
    }
  }

  return {
    singleSelectAnswersList: singleChoiceOptions,
    problemType: dropDownMode ? ProblemTypeKeys.DROPDOWN : ProblemTypeKeys.SINGLESELECT,
  };
}

function getEditorData(markdown) {
  // TODO: Prase heading in this function.
  if (!markdown || !markdown.trim()) {
    return '';
  }
  let description = '';
  const markdownData = markdown.trim().split('\n');
  for (const i in markdownData) {
    const row = markdownData[i];
    if (['{', '(', '[', '=', '|'].indexOf(row.trim()[0]) === -1) {
      const questionTextStart = row.indexOf('>>');
      const questionTextEnd = row.indexOf('<<');
      let questionText = row;
      if (questionTextStart !== -1 && questionTextEnd !== -1 && (questionTextEnd > questionTextStart)) {
        questionText = row.replace('>>', '<p>').replace('<<', '</p>');
      }
      description += `${questionText}\n`;
    } else {
      break;
    }
  }
  return description;
}

function parseMarkdown(markdown) {
  if (!markdown || !markdown.trim()) {
    return {};
  }
  let answers = {};
  let problemType = '';
  const question = getEditorData(markdown);
  const hints = getHints(markdown);
  let data = {question, settings: {hints}};
  let parsedData = getSingleChoiceOptions(markdown);
  if (parsedData.singleSelectAnswersList.length !== 0){
    answers = parsedData.singleSelectAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  parsedData = getMultipleChoiceOptions(markdown);
  if (parsedData.multiSelectAnswersList.length !== 0){
    answers = parsedData.multiSelectAnswersList;
    problemType = parsedData.problemType;
    const groupFeedbackList = parsedData.groupFeedbackList;
    data = {...data, answers, problemType, groupFeedbackList};
  }
  parsedData = getShortAnswerOptions(markdown);
  if (parsedData.shortAnswersList.length !== 0){
    answers = parsedData.shortAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  return data;
}

function getScorringSettings() {
  return {
    selectedAttemptsOption: window.LXCData.max_attempts,
    selectedPointOption: window.LXCData.weight,
  };
}

export {
  getMultipleChoiceOptions,
  getSingleChoiceOptions,
  getEditorData,
  getHints,
  getScorringSettings,
  getShortAnswerOptions,
  parseMarkdown,
};
