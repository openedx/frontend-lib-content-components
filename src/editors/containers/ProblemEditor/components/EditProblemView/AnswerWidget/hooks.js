import { useState, useEffect } from 'react';
import { StrictDict } from '../../../../../utils';
import * as module from './hooks';
import { actions } from '../../../../../data/redux';
import { ProblemTypeKeys } from '../../../../../data/constants/problem';

export const state = StrictDict({
  isFeedbackVisible: (val) => useState(val),
});

export const removeAnswer = ({ answer, dispatch }) => () => {
  dispatch(actions.problem.deleteAnswer({ id: answer.id }));
};

export const setAnswer = ({ answer, hasSingleAnswer, dispatch }) => (payload) => {
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, ...payload }));
};

export const prepareFeedback = (answer) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = module.state.isFeedbackVisible(false);
  useEffect(() => {
    // Show feedback fields if feedback is present
    const isVisible = !!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback;
    setIsFeedbackVisible(isVisible);
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    if (!!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };
  return {
    isFeedbackVisible,
    toggleFeedback,
  };
};

export const isSingleAnswerProblem = (problemType) => (
  problemType === ProblemTypeKeys.DROPDOWN || problemType === ProblemTypeKeys.SINGLESELECT
);

export const useAnswerContainer = ({ answers, updateField }) => {
  useEffect(() => {
    let answerCount = 0;
    answers.forEach(answer => {
      if (answer.correct) {
        answerCount += 1;
      }
    });
    updateField({ correctAnswerCount: answerCount });
  }, []);
};

export default {
  state, removeAnswer, setAnswer, prepareFeedback, isSingleAnswerProblem, useAnswerContainer,
};
