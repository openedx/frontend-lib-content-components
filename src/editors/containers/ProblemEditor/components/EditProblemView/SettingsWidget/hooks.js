import { useState, useEffect } from 'react';

import _ from 'lodash-es';
import * as module from './hooks';
import messages from './messages';
import { ProblemTypeKeys, ShowAnswerTypesKeys } from '../../../../../data/constants/problem';

export const state = {
  showAdvanced: (val) => useState(val),
  cardCollapsed: (val) => useState(val),
  summary: (val) => useState(val),
  showAttempts: (val) => useState(val),
};

export const showAdvancedSettingsCards = () => {
  const [isAdvancedCardsVisible, setIsAdvancedCardsVisible] = module.state.showAdvanced(false);
  return {
    isAdvancedCardsVisible,
    showAdvancedCards: () => setIsAdvancedCardsVisible(true),
  };
};

export const showFullCard = () => {
  const [isCardCollapsed, setIsCardCollapsed] = module.state.cardCollapsed(false);
  return {
    isCardCollapsed,
    toggleCardCollapse: () => setIsCardCollapsed(!isCardCollapsed),
  };
};

export const hintsCardHooks = (hints, updateSettings) => {
  const [summary, setSummary] = module.state.summary({ message: messages.noHintSummary, values: {} });

  useEffect(() => {
    const hintsNumber = hints.length;
    if (hintsNumber === 0) {
      setSummary({ message: messages.noHintSummary, values: {} });
    } else {
      setSummary({ message: messages.hintSummary, values: { hint: hints[0].value, count: (hintsNumber - 1) } });
    }
  }, [hints]);

  const handleAdd = () => {
    let newId = 0;
    if (!_.isEmpty(hints)) {
      newId = Math.max(...hints.map(hint => hint.id)) + 1;
    }
    const hint = { id: newId, value: '' };
    const modifiedHints = [...hints, hint];
    updateSettings({ hints: modifiedHints });
  };

  return {
    summary,
    handleAdd,
  };
};

export const hintsRowHooks = (id, hints, updateSettings) => {
  const handleChange = (event) => {
    const { value } = event.target;
    const modifiedHints = hints.map(hint => {
      if (hint.id === id) {
        return { ...hint, value };
      }
      return hint;
    });
    updateSettings({ hints: modifiedHints });
  };

  const handleDelete = () => {
    const modifiedHints = hints.filter((hint) => (hint.id !== id));
    updateSettings({ hints: modifiedHints });
  };

  return {
    handleChange,
    handleDelete,
  };
};

export const matlabCardHooks = (matLabApiKey, updateSettings) => {
  const [summary, setSummary] = module.state.summary({ message: '', values: {}, intl: false });

  useEffect(() => {
    if (_.isEmpty(matLabApiKey)) {
      setSummary({ message: messages.matlabNoKeySummary, values: {}, intl: true });
    } else {
      setSummary({ message: matLabApiKey, values: {}, intl: false });
    }
  }, [matLabApiKey]);

  const handleChange = (event) => {
    updateSettings({ matLabApiKey: event.target.value });
  };

  return {
    summary,
    handleChange,
  };
};

export const resetCardHooks = (updateSettings) => {
  const setReset = (value) => {
    updateSettings({ showResetButton: value });
  };

  return {
    setResetTrue: () => setReset(true),
    setResetFalse: () => setReset(false),
  };
};

export const scoringCardHooks = (scoring, updateSettings) => {
  const handleMaxAttemptChange = (event) => {
    let unlimitedAttempts = false;
    let attemptNumber = parseInt(event.target.value);
    if (_.isNaN(attemptNumber)) {
      attemptNumber = null;
      unlimitedAttempts = true;
    } else if (attemptNumber < 0) {
      attemptNumber = 0;
    }
    updateSettings({ scoring: { ...scoring, attempts: { number: attemptNumber, unlimited: unlimitedAttempts } } });
  };

  const handleWeightChange = (event) => {
    let weight = parseFloat(event.target.value);
    if (_.isNaN(weight)) {
      weight = 0;
    }
    updateSettings({ scoring: { ...scoring, weight } });
  };

  return {
    handleMaxAttemptChange,
    handleWeightChange,
  };
};

export const showAnswerCardHooks = (showAnswer, updateSettings) => {
  const [showAttempts, setShowAttempts] = module.state.showAttempts(false);
  const numberOfAttemptsChoice = [
    ShowAnswerTypesKeys.AFTER_SOME_NUMBER_OF_ATTEMPTS,
    ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS,
    ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS_OR_CORRECT,
  ];

  useEffect(() => {
    setShowAttempts(_.includes(numberOfAttemptsChoice, showAnswer.on));
  }, [showAttempts]);

  const handleShowAnswerChange = (event) => {
    const { value } = event.target;
    setShowAttempts(_.includes(numberOfAttemptsChoice, value));
    updateSettings({ showAnswer: { ...showAnswer, on: value } });
  };

  const handleAttemptsChange = (event) => {
    let attempts = parseInt(event.target.value);
    if (_.isNaN(attempts)) {
      attempts = 0;
    }
    updateSettings({ showAnswer: { ...showAnswer, afterAttempts: attempts } });
  };

  return {
    handleShowAnswerChange,
    handleAttemptsChange,
    showAttempts,
  };
};

export const timerCardHooks = (updateSettings) => ({
  handleChange: (event) => {
    let time = parseInt(event.target.value);
    if (_.isNaN(time)) {
      time = 0;
    }
    updateSettings({ timeBetween: time });
  },
});

export const typeRowHooks = ({
  answers,
  correctAnswerCount,
  typeKey,
  updateField,
  updateAnswer,
}) => {
  const onClick = () => {
    if (typeKey === ProblemTypeKeys.SINGLESELECT || typeKey === ProblemTypeKeys.DROPDOWN) {
      if (correctAnswerCount > 1) {
        answers.forEach(answer => {
          updateAnswer({ ...answer, correct: false });
        });
      }
    }
    updateField({ problemType: typeKey });
  };
  return {
    onClick,
  };
};

export const confirmSwitchToAdvancedEditor = ({
  switchToAdvancedEditor,
  setConfirmOpen,
}) => {
  switchToAdvancedEditor();
  setConfirmOpen(false);
  window.scrollTo({
    top: 0,
    behavior: 'smooth',
  });
};
