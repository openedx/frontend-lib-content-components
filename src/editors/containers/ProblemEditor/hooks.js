import {
  useRef, useCallback, useState, useEffect,
} from 'react';
import tinyMCEStyles from '../../data/constants/tinyMCEStyles';
import { StrictDict } from '../../utils';
import pluginConfig from '../TextEditor/pluginConfig';
import * as module from './hooks';

export const state = StrictDict({
  refReady: (val) => useState(val),
});

export const parseContentForLabels = ({ editor, updateQuestion }) => {
  let content = editor.getContent();
  const parsedLabels = content.split(/<label>|<\/label>/gm);
  let updatedContent;
  parsedLabels.forEach((label, i) => {
    let updatedLabel = label;
    if (!label.startsWith('<') && !label.endsWith('>')) {
      let previousLabel = parsedLabels[i - 1];
      let nextLabel = parsedLabels[i + 1];
      if (!previousLabel.endsWith('<p>')) {
        previousLabel = `${previousLabel}</p><p>`;
        updatedContent = content.replace(parsedLabels[i - 1], previousLabel);
        content = updatedContent;
      }
      if (previousLabel.endsWith('</p>') && !label.startWith('<p>')) {
        updatedLabel = `<p>${label}`;
        updatedContent = content.replace(label, updatedLabel);
        content = updatedContent;
      }
      if (!nextLabel.startsWith('</p>')) {
        nextLabel = `</p><p>${nextLabel}`;
        updatedContent = content.replace(parsedLabels[i + 1], nextLabel);
        content = updatedContent;
      }
    }
  });
  updateQuestion(content);
};

export const setupCustomBehavior = ({ updateQuestion }) => (editor) => {
  // add a custom simple inline label formatter.
  const toggleLabelFormatting = () => {
    editor.execCommand('mceToggleFormat', false, 'label');
  };
  editor.ui.registry.addIcon('textToSpeech',
    '<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M3.5 22C3.08333 22 2.72917 21.8542 2.4375 21.5625C2.14583 21.2708 2 20.9167 2 20.5V3.5C2 3.08333 2.14583 2.72917 2.4375 2.4375C2.72917 2.14583 3.08333 2 3.5 2H13L11.5 3.5H3.5V20.5H15.5V17H17V20.5C17 20.9167 16.8542 21.2708 16.5625 21.5625C16.2708 21.8542 15.9167 22 15.5 22H3.5ZM6 17.75V16.25H13V17.75H6ZM6 14.75V13.25H11V14.75H6ZM15.5 15L11.5 11H8V6H11.5L15.5 2V15ZM17 12.7V4.05C17.9333 4.4 18.6667 5.01667 19.2 5.9C19.7333 6.78333 20 7.65 20 8.5C20 9.35 19.7083 10.1917 19.125 11.025C18.5417 11.8583 17.8333 12.4167 17 12.7ZM17 16.25V14.7C18.1667 14.2833 19.2083 13.5333 20.125 12.45C21.0417 11.3667 21.5 10.05 21.5 8.5C21.5 6.95 21.0417 5.63333 20.125 4.55C19.2083 3.46667 18.1667 2.71667 17 2.3V0.75C18.7 1.2 20.125 2.1375 21.275 3.5625C22.425 4.9875 23 6.63333 23 8.5C23 10.3667 22.425 12.0125 21.275 13.4375C20.125 14.8625 18.7 15.8 17 16.25Z" fill="black"/></svg>');
  editor.ui.registry.addButton('customLabelButton', {
    icon: 'textToSpeech',
    text: 'Label',
    tooltip: 'Apply a "Question" label to specific text, recognized by screen readers. Recommended to improve accessibility.',
    onAction: toggleLabelFormatting,
  });
  editor.on('blur', () => {
    module.parseContentForLabels({
      editor,
      updateQuestion,
    });
  });
};

export const problemEditorConfig = ({
  setEditorRef,
  question,
  updateQuestion,
}) => ({
  onInit: (evt, editor) => {
    setEditorRef(editor);
  },
  initialValue: question || '',
  init: {
    skin: false,
    content_css: false,
    content_style: tinyMCEStyles,
    menubar: false,
    branding: false,
    min_height: 150,
    placeholder: 'Enter your question',
    formats: { label: { inline: 'label' } },
    setup: module.setupCustomBehavior({ updateQuestion }),
    toolbar: `${pluginConfig().toolbar} | customLabelButton`,
    plugins: 'autoresize',
  },
});

export const prepareEditorRef = () => {
  const editorRef = useRef(null);
  const setEditorRef = useCallback((ref) => {
    editorRef.current = ref;
  }, []);
  const [refReady, setRefReady] = module.state.refReady(false);
  useEffect(() => setRefReady(true), [setRefReady]);
  return { editorRef, refReady, setEditorRef };
};
