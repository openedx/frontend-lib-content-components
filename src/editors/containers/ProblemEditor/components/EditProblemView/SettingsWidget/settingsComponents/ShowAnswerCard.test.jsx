import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import { formatMessage } from '../../../../../../../testUtils';
import { selectors } from '../../../../../../data/redux';
import { ShowAnswerCard, mapStateToProps, mapDispatchToProps } from './ShowAnswerCard';
import { useAnswerSettings } from '../hooks';

jest.mock('../hooks', () => ({
  useAnswerSettings: jest.fn(),
}));

jest.mock('../../../../../../data/redux', () => ({
  selectors: {
    app: {
      studioEndpointUrl: jest.fn(state => ({ studioEndpointUrl: state })),
      learningContextId: jest.fn(state => ({ learningContextId: state })),
      isLibrary: jest.fn(state => ({ isLibrary: state })),
    },
  },
  thunkActions: {
    video: jest.fn(),
  },
}));

describe('ShowAnswerCard', () => {
  const showAnswer = {
    on: 'after_attempts',
    afterAttempts: 5,
    updateSettings: jest.fn().mockName('args.updateSettings'),
    intl: { formatMessage },
  };
  const props = {
    showAnswer,
    defaultValue: 'finished',
    // injected
    intl: { formatMessage },
    // redux
    studioEndpointUrl: 'SoMEeNDpOinT',
    learningContextId: 'sOMEcouRseId',
  };

  const useAnswerSettingsProps = {
    handleShowAnswerChange: jest.fn().mockName('useAnswerSettings.handleShowAnswerChange'),
    handleAttemptsChange: jest.fn().mockName('useAnswerSettings.handleAttemptsChange'),
  };

  useAnswerSettings.mockReturnValue(useAnswerSettingsProps);

  describe('behavior', () => {
    it(' calls useAnswerSettings when initialized', () => {
      shallow(<ShowAnswerCard {...props} />);
      expect(useAnswerSettings).toHaveBeenCalledWith(showAnswer, props.updateSettings);
    });
  });

  describe('snapshot', () => {
    test('snapshot: show answer setting card', () => {
      expect(shallow(<ShowAnswerCard {...props} />).snapshot).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('studioEndpointUrl from app.studioEndpointUrl', () => {
      expect(
        mapStateToProps(testState).studioEndpointUrl,
      ).toEqual(selectors.app.studioEndpointUrl(testState));
    });
    test('learningContextId from app.learningContextId', () => {
      expect(
        mapStateToProps(testState).learningContextId,
      ).toEqual(selectors.app.learningContextId(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    test('equal an empty object', () => {
      expect(mapDispatchToProps).toEqual({});
    });
  });
});
