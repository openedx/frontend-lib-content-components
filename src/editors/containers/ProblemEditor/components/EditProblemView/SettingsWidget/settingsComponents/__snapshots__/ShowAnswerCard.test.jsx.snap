// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowAnswerCard snapshot snapshot: show answer setting card 1`] = `
<SettingsOption
  className=""
  extraSections={Array []}
  hasExpandableTextArea={false}
  summary="After Some Number of Attempts"
  title="Show answer"
>
  <Fragment>
    <div
      className="pb-2"
    >
      <span>
        <FormattedMessage
          defaultMessage="Define when learners can see the correct answer."
          description="Show Answer settings card text"
          id="authoring.problemeditor.settings.showAnswer.text"
        />
      </span>
    </div>
    <div
      className="pb-4"
    >
      <Hyperlink
        destination="SoMEeNDpOinT/settings/advanced/sOMEcouRseId#showanswer"
        target="_blank"
      >
        <FormattedMessage
          defaultMessage="Set a default value in advanced settings"
          description="Advanced settings link text"
          id="authoring.problemeditor.settings.advancedSettingLink.text"
        />
      </Hyperlink>
    </div>
    <Form.Group
      className="pb-0 mb-0"
    >
      <Form.Control
        as="select"
        onChange={[MockFunction useAnswerSettings.handleShowAnswerChange]}
        value="after_attempts"
      >
        <option
          key="always"
          value="always"
        >
          Always
        </option>
        <option
          key="answered"
          value="answered"
        >
          Answered
        </option>
        <option
          key="attempted"
          value="attempted"
        >
          Attempted or Past Due
        </option>
        <option
          key="closed"
          value="closed"
        >
          Closed
        </option>
        <option
          key="finished"
          value="finished"
        >
          Finished (Default)
        </option>
        <option
          key="correct_or_past_due"
          value="correct_or_past_due"
        >
          Correct or Past Due
        </option>
        <option
          key="past_due"
          value="past_due"
        >
          Past Due
        </option>
        <option
          key="never"
          value="never"
        >
          Never
        </option>
        <option
          key="after_attempts"
          value="after_attempts"
        >
          After Some Number of Attempts
        </option>
        <option
          key="after_all_attempts"
          value="after_all_attempts"
        >
          After All Attempts
        </option>
        <option
          key="after_all_attempts_or_correct"
          value="after_all_attempts_or_correct"
        >
          After All Attempts or Correct
        </option>
        <option
          key="attempted_no_past_due"
          value="attempted_no_past_due"
        >
          Attempted
        </option>
      </Form.Control>
    </Form.Group>
  </Fragment>
</SettingsOption>
`;
